src/topology/continuous_function/stone_weierstrass.lean

```diff
@@ -0,0 +1,353 @@
+  apply eq_top_iff.mpr,
+  rintros f -,
+  refine filter.frequently.mem_closure
+    ((filter.has_basis.frequently_iff metric.nhds_basis_ball).mpr (λ ε pos, _)),
+  simp only [exists_prop, metric.mem_ball],
+
+  -- It will be helpful to assume `X` is nonempty later,
+  -- so we get that out of the way here.
+  by_cases nX : nonempty X,
+  swap,
+  exact ⟨nA.some, (dist_lt_iff _ _ pos).mpr (λ x, false.elim (nX ⟨x⟩)), nA.some_spec⟩,
+
+  /-
+  The strategy now is to pick a family of continuous functions `g x y` in `A`
+  with the property that `g x y x = f x` and `g x y y = f y`
+  (this is immediate from `h : separates_points_strongly`)
+  then use continuity to see that `g x y` is close to `f` near both `x` and `y`,
+  and finally using compactness to produce the desired function `h`
+  as a maximum over finitely many `x` of a minimum over finitely many `y` of the `g x y`.
+  -/
+  dsimp [set.separates_points_strongly] at sep,
+
+  let g : Π x y, L := λ x y, (sep f x y).some,
+  let w₁ : ∀ x y, g x y x = f x := λ x y, (sep f x y).some_spec.1,
+  let w₂ : ∀ x y, g x y y = f y := λ x y, (sep f x y).some_spec.2,
+
+  -- For each `x y`, we define `U x y` to be `{ z | f z - ε < g x y z }`,
+  -- and observe this is a neighbourhood of `y`.
+  let U : Π x y, set X := λ x y, { z | f z - ε < g x y z },
+  have U_nhd_y : ∀ x y, U x y ∈ 𝓝 y,
+  { intros x y,
+    refine mem_nhds_sets _ _,
+    { rw [show U x y = (f - g x y : C(X, ℝ)) ⁻¹' set.Iio ε, { ext z, simp [sub_lt], }],
+      exact is_open.preimage (by continuity) is_open_Iio, },
+    { rw [set.mem_set_of_eq, w₂],
+      exact sub_lt_self _ pos, }, },
+
+  -- Fixing `x` for a moment, we have a family of functions `λ y, g x y`
+  -- which on different patches (the `U x y`) are greater than `f z - ε`.
+  -- Taking the supremum of these functions
+  -- indexed by a finite collection of patches which cover `X`
+  -- will give us an element of `A` that is globally greater than `f z - ε`.
+
+  -- Since `X` is compact, for every `x` there is some finset `ys t`
+  -- so the union of the `U x y` for `y ∈ ys x` still covers everything.
+  let ys : Π x, finset X := λ x, (compact_space.elim_nhds_subcover (U x) (U_nhd_y x)).some,
+  let ys_w : ∀ x, (⋃ y ∈ ys x, U x y) = ⊤ :=
+    λ x, (compact_space.elim_nhds_subcover (U x) (U_nhd_y x)).some_spec,
+  have ys_nonempty : ∀ x, (ys x).nonempty :=
+    λ x, set.nonempty_of_union_eq_top_of_nonempty _ _ nX (ys_w x),
+
+  -- Thus for each `x` we have the desired `h x : A` so `f z - ε < h x z` everywhere.
+  let h : Π x, L := λ x,
+    ⟨(ys x).sup' (ys_nonempty x) (λ y, (g x y : C(X, ℝ))),
+      finset.sup'_mem _ sup_mem _ _ _ (λ y _, (g x y).2)⟩,
+  have lt_h : ∀ x z, f z - ε < h x z,
+  { intros x z,
+    obtain ⟨y, ym, zm⟩ := set.exists_set_mem_of_union_eq_top _ _ (ys_w x) z,
+    dsimp [h],
+    simp only [finset.lt_sup'_iff, continuous_map.sup'_apply],
+    exact ⟨y, ym, zm⟩, },
+
+  -- For each `x y`, we define `V x y` to be `{ z | g x y z < f z + ε }`,
```

Comment:
I think that the proof is too complicated in the next 20 lines. You should completely drop the functions `g`, and define everything in terms of `h`. I.e., define directly `W x = {z | h x z < f z + ε}`, and notice that this is an open neighborhood of `x` as `h x` is continuous and equal to `f x` at `x` .

Reply:
@sgouezel, I've made this proposed changed in https://github.com/leanprover-community/mathlib/pull/7305/commits/a7d3f81aeda657fa4d6084509af6aa08129b7ea8.

However, I've ended up not entirely convinced by it; certainly as a maths-proof it's an improvement. I'll make separate comments asking for advice in the places I'm not happy with.